### DISCUSSION TITLE
Seem to continually get bad connections after hours of inactivity on deployed python app

### DISCUSSION THREAD

**Discussion by thedonmon** [psycopg/psycopg#444] [Q&A]
Hello, I have a discord bot running that is deployed to fly.io. I am using psycopg3 and seem to continually get bad connections in my connection pool after hours of inactivity. So then when I ping the bot a command, the whole application crashes from this error: `discarding broken connection: <psycopg.Connection [BAD] at 0x7f997b5b6bd0>`

Below is a sample call please let me know if there is a better way to do this?
```python
def executeF1(command, *values):
    try:
        pool.check()
        with pool.connection() as conn:
            with conn.cursor() as cur:
                cur.execute(command, tuple(values))
                result = cur.fetchone()
    except:
        print('Could not complete query "{}"'.format(command))
        print(traceback.format_exc())
        result = None
    finally:
        pool.check()
    return result
   ```

It would be great to get some best practice on this issue. When converting to async I ran into even more issues. I am using a pool because I do not want to instantiate a new connection everytime.

**Comment by dvarrazzo** [comment]
Most likely they are closing the connection. I have a similar problem with an app which to postgres via haproxy to listen for notifications, and haproxy closes the connection every hour or so.

`discarding broken connection` is not an error, it is a warning in the pool and it's normal to get it at the end of the `pool.connection()` block if the connection is closed. The pool will immediately try to create a new connection, so it's not what makes your program crash.

The docs have info about [checking pool connections](https://www.psycopg.org/psycopg3/docs/advanced/pool.html#connection-quality) and [checking for disconnections](https://www.psycopg.org/psycopg3/docs/advanced/async.html#disconnections): you can take a look to some of these pointers. Another way I've experimented lately, which encapsulate the waiting, is to use [`connection.notifies()`](https://www.psycopg.org/psycopg3/docs/advanced/async.html#asynchronous-notifications), which should terminate as soon as the connection is closed:

```
In [1]: import psycopg
In [2]: conn = psycopg.connect("")
In [3]: print(conn.info.backend_pid)
359917
In [4]: with conn.notifies():
   ...:     pass
```

in another shell:

```
$ psql -c "select pg_terminate_backend(359917)"
┌──────────────────────┐
│ pg_terminate_backend │
├──────────────────────┤
│ t                    │
└──────────────────────┘
(1 row)
```

in the python shell you immediately receive:

```
OperationalError: consuming input failed: server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
```

if you have a connection pool, you can use a connection to detect for disconnection and one for work. when a disconnection is detected you can cancel worker and reconnect using a fresh connection.

**Reply by thedonmon** [reply]
> Most likely they are closing the connection. I have a similar problem with an app which to postgres via haproxy to listen for notifications, and haproxy closes the connection every hour or so.
> 
> `discarding broken connection` is not an error, it is a warning in the pool and it's normal to get it at the end of the `pool.connection()` block if the connection is closed. The pool will immediately try to create a new connection, so it's not what makes your program crash.
> 
> The docs have info about [checking pool connections](https://www.psycopg.org/psycopg3/docs/advanced/pool.html#connection-quality) and [checking for disconnections](https://www.psycopg.org/psycopg3/docs/advanced/async.html#disconnections): you can take a look to some of these pointers. Another way I've experimented lately, which encapsulate the waiting, is to use [`connection.notifies()`](https://www.psycopg.org/psycopg3/docs/advanced/async.html#asynchronous-notifications), which should terminate as soon as the connection is closed:
> 
> ```
> In [1]: import psycopg
> In [2]: conn = psycopg.connect("")
> In [3]: print(conn.info.backend_pid)
> 359917
> In [4]: with conn.notifies():
>    ...:     pass
> ```
> 
> in another shell:
> 
> ```
> $ psql -c "select pg_terminate_backend(359917)"
> ┌──────────────────────┐
> │ pg_terminate_backend │
> ├──────────────────────┤
> │ t                    │
> └──────────────────────┘
> (1 row)
> ```
> 
> in the python shell you immediately receive:
> 
> ```
> OperationalError: consuming input failed: server closed the connection unexpectedly
> 	This probably means the server terminated abnormally
> 	before or while processing the request.
> ```
> 
> if you have a connection pool, you can use a connection to detect for disconnection and one for work. when a disconnection is detected you can cancel worker and reconnect using a fresh connection.

Ok so does that code block generally look ok? I just had a problem with insert many and somehow this seemed to be the fix and I have no idea why...
```python
def multiexec(command, valueset):
    try:
        print(f"in multiexec: {command}")
        pool.check()
        conn = pool.getconn(timeout=30)
        conn._check_connection_ok()
        conn._prepared.clear()
        if conn.broken or conn.TransactionStatus == 0:
            print("conn broken")
            pool.putconn(conn)
            conn = Connection(DATABASE_URL)
        conn.autocommit = True
        with conn.cursor() as cur:
            result = cur.executemany(command, valueset)
        conn.commit()
        pool.putconn(conn)
    except:
        print('Could not complete query "{}"'.format(command))
        print(traceback.format_exc())
        result = None
        pool.check()
    finally:
        pool.check()
    return result
```

I will take a look at setting up the notifies piece but not sure exactly how that works. 

Would I be able to do something like this in a seperate module?

```python
import asyncio
from db.db import pool
ev = asyncio.Event()
conn = pool.getconn()
loop = asyncio.get_event_loop()
loop.add_reader(conn.fileno(), ev.set)

async def poll_conn():
    while True:
        try:
            await asyncio.wait_for(ev.wait(), 60.0)
        except asyncio.TimeoutError:
            continue  # No FD activity detected in one minute

    # Activity detected. Is the connection still ok?
        try:
            await conn.execute("SELECT 1")
        except psycopg.OperationalError:
            print(f"Connection lost. Reconnecting...")
```

**Reply by thedonmon** [reply]
then call in my bot.py `asyncio.run(polling.poll_conn())`

**Comment by parity3** [comment]
Some pointers from the ops side of things:

Make sure that the service is not deployed using virtual ips (use dns-rr or host networking instead). This bit me on docker swarm. You may also want to force a default for TCP keepalives via:

```yml
  sysctls:
    - net.ipv4.tcp_keepalive_time=30
    - net.ipv4.tcp_keepalive_intvl=15
    - net.ipv4.tcp_keepalive_probes=2
```

The virtual IP method seemed to hard-drop the connections without warning or sending any resets to either side, and they also seemed to have timeouts that would break connections even if no containers changed. After adding those settings my issues with disconnects disappeared.
